From 23f30fb5a98a6c464ff8897177e5f97267850cbd Mon Sep 17 00:00:00 2001
From: nikolay serdjuk <nikolay.y.serdjuk@intel.com>
Date: Tue, 24 Dec 2013 12:10:02 +0700
Subject: Dalvik: A fix for SegFault which occurs in MethodContext::handleConstants

BZ: 160554

A comparison against past-the-end element in the map container was added.

Change-Id: I40f78587c349cff45df913134a9ff262cfaf2137
Category: device enablement
Domain: AOSP.Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, need rework
Signed-off-by: nikolay serdjuk <nikolay.y.serdjuk@intel.com>
---
 vm/compiler/MethodContext.cpp |   10 +++++++++-
 1 files changed, 9 insertions(+), 1 deletions(-)

diff --git a/vm/compiler/MethodContext.cpp b/vm/compiler/MethodContext.cpp
index 95f6997..75344d9 100644
--- a/vm/compiler/MethodContext.cpp
+++ b/vm/compiler/MethodContext.cpp
@@ -500,7 +500,15 @@ bool MethodContext::handleConstants (CompilationUnit *cUnit)
             int VR = it->first;
 
             //Get constant index of this VR
-            const std::vector<ConstOffset>& offsets = vRConstMap.find (VR)->second;
+            const std::map<int, std::vector<ConstOffset> >::const_iterator mapIterator = vRConstMap.find (VR);
+
+            //Fail if the VR is not found
+            if (mapIterator == vRConstMap.end ())
+            {
+                return false;
+            }
+
+            const std::vector<ConstOffset>& offsets = mapIterator->second;
 
             //Paranoid: The VR should have at least one const value
             if (offsets.size() < 1)
-- 
1.7.4.1

